feat(incidents): add ability to resolve incidents

blestab opened the pull request:
Fixes #2078.
Fixes #2218.

**Changes proposed in this pull request:**

***Incident View***
- Add new user permission _CompleteIncident_
- Add new incident property _CompletedOn_ 
- Add new incident status _completed_
- Add Complete Incident button on ViewIncident screen
- Add Completed On display information on the ViewIncident screen for a resolved incident
- Display _Reported By_ username correctly
- Unit Tests 
-- Show Complete Incident button when incident is not completed
-- Show Complete Incident button when user has the access
-- Do not show the Complete button when incident is already completed
-- Do not show the Complete button if user does not have the access
-- Show the Completed On details if the incident is completed
-- Do not show the Completed On details if the incident is not completed
-- Extract username utility unit test

***Incidents View***
- Display _Reported By_ username correctly
- Add _completed_ status to list of filter options

***Incidents Utilities***
- Utility to extract username from the couchdb user id string

blestab commented:
@jackcmeyer @kumikokashii @tehKapa sorry to bug, may i have some guidance/suggestions regarding items **[Q1]** and **[Q2]** in my draft PR.

fox1t commented:
As for Q1 you are trying to modify in place an object that is not [extensible](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cant_define_property_object_not_extensible)
try to write it like
```typescript
  const completedIncident = await IncidentRepository.saveOrUpdate({
    ...incidentToComplete,
    completedOn: new Date(Date.now().valueOf()).toISOString(),
    status: 'completed',
  })
```
You cannot modify objects in place: you have to "generate" a new object and copy props from the old one.

Q2 seems that it need a fix. You can open an another PR if you want to work on it :)

blestab commented:
Thanks @fox1t i could have sworn that i tried this before and it refused to work 😇 , works this time around though so that one is now fixed, thanks.

I have also fixed Q2 as part of this PR so will update the description.

Just need to add some tests then i can remove it from draft status

blestab commented:
Hi all, 
PR is now ready for review

jackcmeyer commented on src/incidents/util/extractUsername.ts:
```diff
@@ -0,0 +1,2 @@
+export const extractUsername = (username: string) =>
```
this util can probably used in several different spots. To indicate that, I think this is better positioned in `src/shared/utils` folder

jackcmeyer commented on src/incidents/incident-slice.ts:
```diff
@@ -120,4 +124,23 @@ export const reportIncident = (
   }
 }
 
+export const completeIncident = (
+  incidentToComplete: Incident,
+  onSuccess?: (incidentToComplete: Incident) => void,
+): AppThunk => async (dispatch) => {
+  dispatch(completeIncidentStart())
+
+  const completedIncident = await IncidentRepository.saveOrUpdate({
+    ...incidentToComplete,
+    completedOn: new Date(Date.now().valueOf()).toISOString(),
+    status: 'completed',
+  })
+
+  dispatch(completeIncidentSuccess(completedIncident))
+
+  if (onSuccess) {
+    onSuccess(completedIncident)
+  }
+}
```
there should be a test for this functionality

blestab replied:
I have added an 'on resolve' test into the ViewIncident test file, will that be ok?

blestab replied:
Actually, just had a look at some lab module examples, I'll go ahead write tests in the incident-slice as well.

jackcmeyer commented on src/__tests__/incidents/util/extract-username-util.test.ts:
```diff
@@ -0,0 +1,6 @@
+import { extractUsername } from '../../../incidents/util/extractUsername'
+
+it('should extract the string after the last : in a given string', () => {
```
the file name for this test should mirror the file name for the implementation.

jackcmeyer replied:
a convention the project has been following is that all tests are wrapped in a top-level describe. 

For this case it might be something like

```
describe('extract username util', () => {
	it('...', () => {})
})
```

blestab commented:
Thanks @jackcmeyer 
I have made code changes as suggested.
